// @forestadmin/datasource-toolkit/dist/src/validation/rules.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MAP_ALLOWED_TYPES_FOR_OPERATOR = exports.MAP_ALLOWED_TYPES_FOR_COLUMN_TYPE = exports.MAP_ALLOWED_OPERATORS_FOR_COLUMN_TYPE = void 0;
const types_1 = require("./types");
const BASE_OPERATORS = ['Blank', 'Equal', 'Missing', 'NotEqual', 'Present'];
const ARRAY_OPERATORS = ['In', 'NotIn', 'IncludesAll'];
const BASE_DATEONLY_OPERATORS = [
    'Today',
    'Yesterday',
    'PreviousXDaysToDate',
    'PreviousWeek',
    'PreviousWeekToDate',
    'PreviousMonth',
    'PreviousMonthToDate',
    'PreviousQuarter',
    'PreviousQuarterToDate',
    'PreviousYear',
    'PreviousYearToDate',
    'Past',
    'Future',
    'PreviousXDays',
    'Before',
    'After',
];
exports.MAP_ALLOWED_OPERATORS_FOR_COLUMN_TYPE = Object.freeze({
    String: [
        ...BASE_OPERATORS,
        ...ARRAY_OPERATORS,
        'Contains',
        'NotContains',
        'EndsWith',
        'StartsWith',
        'LongerThan',
        'ShorterThan',
        'Like',
        'ILike',
        'IContains',
        'IEndsWith',
        'IStartsWith',
    ],
    Number: [...BASE_OPERATORS, ...ARRAY_OPERATORS, 'GreaterThan', 'LessThan'],
    Dateonly: [...BASE_OPERATORS, ...BASE_DATEONLY_OPERATORS],
    Date: [...BASE_OPERATORS, ...BASE_DATEONLY_OPERATORS, 'BeforeXHoursAgo', 'AfterXHoursAgo'],
    Timeonly: [...BASE_OPERATORS, 'LessThan', 'GreaterThan'],
    Enum: [...BASE_OPERATORS, ...ARRAY_OPERATORS],
    Json: ['Blank', 'Missing', 'Present'],
    Boolean: BASE_OPERATORS,
    Point: BASE_OPERATORS,
    Uuid: [...BASE_OPERATORS, ...ARRAY_OPERATORS],
});
exports.MAP_ALLOWED_TYPES_FOR_COLUMN_TYPE = Object.freeze({
    String: ['String', types_1.ValidationTypesArray.String, types_1.ValidationPrimaryTypes.Null],
    Number: ['Number', types_1.ValidationTypesArray.Number, types_1.ValidationPrimaryTypes.Null],
    Boolean: ['Boolean', types_1.ValidationTypesArray.Boolean, types_1.ValidationPrimaryTypes.Null],
    Enum: ['Enum', types_1.ValidationTypesArray.Enum, types_1.ValidationPrimaryTypes.Null],
    Date: ['Date', types_1.ValidationPrimaryTypes.Null],
    Dateonly: ['Dateonly', types_1.ValidationPrimaryTypes.Null],
    Json: ['Json', types_1.ValidationPrimaryTypes.Null],
    Point: ['Point', types_1.ValidationPrimaryTypes.Null],
    Timeonly: ['Timeonly', types_1.ValidationPrimaryTypes.Null],
    Uuid: ['Uuid', types_1.ValidationTypesArray.Uuid, types_1.ValidationPrimaryTypes.Null],
});
function computeAllowedTypesForOperators() {
    return Object.keys(exports.MAP_ALLOWED_OPERATORS_FOR_COLUMN_TYPE).reduce((mapMemo, type) => {
        const allowedOperators = exports.MAP_ALLOWED_OPERATORS_FOR_COLUMN_TYPE[type];
        allowedOperators.forEach(operator => {
            if (mapMemo[operator]) {
                mapMemo[operator].push(type);
            }
            else {
                mapMemo[operator] = [type];
            }
        });
        return mapMemo;
    }, {});
}
const NO_TYPES_ALLOWED = [types_1.ValidationPrimaryTypes.Null];
exports.MAP_ALLOWED_TYPES_FOR_OPERATOR = Object.freeze({
    ...computeAllowedTypesForOperators(),
    In: Object.values(types_1.ValidationTypesArray),
    NotIn: Object.values(types_1.ValidationTypesArray),
    IncludesAll: Object.values(types_1.ValidationTypesArray),
    Blank: NO_TYPES_ALLOWED,
    Missing: NO_TYPES_ALLOWED,
    Present: NO_TYPES_ALLOWED,
    Yesterday: NO_TYPES_ALLOWED,
    Today: NO_TYPES_ALLOWED,
    PreviousQuarter: NO_TYPES_ALLOWED,
    PreviousYear: NO_TYPES_ALLOWED,
    PreviousMonth: NO_TYPES_ALLOWED,
    PreviousWeek: NO_TYPES_ALLOWED,
    Past: NO_TYPES_ALLOWED,
    Future: NO_TYPES_ALLOWED,
    PreviousWeekToDate: NO_TYPES_ALLOWED,
    PreviousMonthToDate: NO_TYPES_ALLOWED,
    PreviousQuarterToDate: NO_TYPES_ALLOWED,
    PreviousYearToDate: NO_TYPES_ALLOWED,
    PreviousXDaysToDate: ['Number'],
    PreviousXDays: ['Number'],
    BeforeXHoursAgo: ['Number'],
    AfterXHoursAgo: ['Number'],
    LongerThan: ['Number'],
    ShorterThan: ['Number'],
});
