// @forestadmin/datasource-toolkit/dist/src/decorators/validation/collection.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const errors_1 = require("../../errors");
const collection_decorator_1 = __importDefault(require("../collection-decorator"));
const factory_1 = __importDefault(require("../../interfaces/query/condition-tree/factory"));
const condition_tree_1 = __importDefault(require("../../validation/condition-tree"));
const field_1 = __importDefault(require("../../validation/field"));
class ValidationDecorator extends collection_decorator_1.default {
    constructor() {
        super(...arguments);
        this.validation = {};
    }
    addValidation(name, validation) {
        var _a;
        field_1.default.validate(this, name);
        const field = this.childCollection.schema.fields[name];
        if (!field)
            throw new Error('Cannot add validators on a relation, use the foreign key instead');
        if (field.isReadOnly)
            throw new Error('Cannot add validators on a readonly field');
        (_a = this.validation)[name] ?? (_a[name] = []);
        this.validation[name].push(validation);
        this.markSchemaAsDirty();
    }
    async create(caller, data) {
        for (const record of data)
            this.validate(record, caller.timezone, true);
        return super.create(caller, data);
    }
    async update(caller, filter, patch) {
        this.validate(patch, caller.timezone, false);
        return super.update(caller, filter, patch);
    }
    refineSchema(subSchema) {
        const schema = { ...subSchema, fields: { ...subSchema.fields } };
        for (const [name, rules] of Object.entries(this.validation)) {
            const field = { ...schema.fields[name] };
            field.validation = ValidationDecorator.deduplicate([...(field.validation ?? []), ...rules]);
            schema.fields[name] = field;
        }
        return schema;
    }
    validate(record, timezone, allFields) {
        for (const [name, field] of Object.entries(this.schema.fields)) {
            if (field.type === 'Column' && (allFields || record[name] !== undefined)) {
                // When setting a field to null, we skip all validators but "Present"
                let rules = field.validation ?? [];
                if (record[name] === null)
                    rules = rules.filter(r => r.operator === 'Present');
                for (const validator of rules) {
                    const rawLeaf = { field: name, ...validator };
                    const tree = factory_1.default.fromPlainObject(rawLeaf);
                    condition_tree_1.default.validate(tree, this);
                    if (!tree.match(record, this, timezone)) {
                        const message = `'${name}' failed validation rule:`;
                        const rule = validator.value !== undefined
                            ? `${validator.operator}(${validator.value})`
                            : `${validator.operator}`;
                        throw new errors_1.ValidationError(`${message} '${rule}'`);
                    }
                }
            }
        }
    }
    /**
     * Deduplicate rules which the frontend understand
     * We ignore other rules as duplications are not an issue within the agent
     */
    static deduplicate(rules) {
        var _a;
        const values = {};
        for (const rule of rules) {
            values[_a = rule.operator] ?? (values[_a] = []);
            values[rule.operator].push(rule);
        }
        // Remove duplicate "Present"
        while (values.Present?.length > 1)
            values.Present.pop();
        // Merge duplicate 'GreaterThan', 'After' and 'LongerThan' (keep the max value)
        for (const operator of ['GreaterThan', 'After', 'LongerThan']) {
            while (values[operator]?.length > 1) {
                const last = values[operator].pop();
                values[operator][0] = {
                    operator,
                    value: ValidationDecorator.max(last.value, values[operator][0].value),
                };
            }
        }
        // Merge duplicate 'LessThan', 'Before' and 'ShorterThan' (keep the min value)
        for (const operator of ['LessThan', 'Before', 'ShorterThan']) {
            while (values[operator]?.length > 1) {
                const last = values[operator].pop();
                values[operator][0] = {
                    operator,
                    value: ValidationDecorator.min(last.value, values[operator][0].value),
                };
            }
        }
        return Object.values(values).reduce((memo, r) => [...memo, ...r], []);
    }
    static min(valueA, valueB) {
        return valueA < valueB ? valueA : valueB;
    }
    static max(valueA, valueB) {
        return valueA < valueB ? valueB : valueA;
    }
}
exports.default = ValidationDecorator;
